### PR TITLE
CAS: Rewrite database to avoid storing hashes redundantly with each reference

### PR DESCRIPTION
**Pull request by dexonsmith**
Lots of commits in here; I should squash most of them probably.

This required a series of major refactorings... hopefully the new state will be easier to maintain.

One neat trick in the `OnDiskCAS` is that there's a single `mmap` region where all the hashes should live. This allows avoiding subsequent lookups of the same ID:
```
OnDiskHashMappedTrie::const_pointer
OnDiskCAS::getInternalIndexPointer(CASID ID) const {
  // Lookup the hash by pointer value.
  //
  // FIXME: Can make this more robust by changing CASID to a CASDB and an
  // integer, where the integer is an offset into the trie.
  if (OnDiskHashMappedTrie::const_pointer P =
          Index.recoverFromHashPointer(ID.getHash().data())) {
    assert(P->Hash == ID.getHash());
    return P;
  }

  // Fallback to a normal lookup.
  return Index.find(ID.getHash());
}
```
I.e., instead of doing the usual `Index.find()` to lookup the hash up in the hash-mapped-trie, we first check if the hash is pointing inside the `memory_mapped_region` for the `Index` file. This is likely, since every CASID returned by the `OnDiskCAS` is memoized there (including results of `parseCASID()`).

The same basic design is being used for the `InMemoryCAS` but we can't quite do the same trick, since it's not trivial to detect whether it's in one of the slabs of the bump-ptr-allocator. I think the right thing is to change `CASID` to something like:
```
class CASID {
public:
  ArrayRef<uint8_t> getHash() const { return DB->getHash(ID); }
  bool operator==(const CASID &RHS) const {
    if (DB == RHS.DB)
      return ID == RHS.ID;
    return DB->equals(ID, RHS);
  }
private:
  CASDB *DB;
  uint64_t ID;
};
```
Then the in-memory CAS can also avoid the lookups, by storing a pointer value as `ID`.

Anyway, I'm not looking for specific review here—probably too big of a change to review easily—but if you see something please let me know.

### CODE REVIEW

**Comment by dexonsmith** [comment]
@swift-ci test

File: llvm/lib/CAS/BuiltinCAS.cpp [hunk 1]
```diff
@@ -353,784 +287,2929 @@ Error BuiltinCAS::printCASID(raw_ostream &OS, CASID ID) {
...
-CASID BuiltinCAS::ObjectContentReference::getReference(size_t I) const {
-  assert(I < getNumReferences() && "Reference out of range");
-  return CASID(
-      bufferAsRawHash(ReferenceBlock.substr(I * NumHashBytes, NumHashBytes)));
+Expected<NodeRef> BuiltinCAS::createNode(ArrayRef<CASID> References,
+                                         ArrayRef<char> Data) {
+  BuiltinNodeHasher<HasherT> Hasher;
+  Hasher.start(References.size());
+  for (const CASID &ID : References)
+    Hasher.updateRef(ID.getHash());
```
Comment (cachemeifyoucan) [inline llvm/lib/CAS/BuiltinCAS.cpp:358]: `ID.getHash()` returns the CAS internal representation right? Should this be based on something more stable like the hash value with the actual hash algorithm?

For example, if you have InMemoryCAS and OnDiskCAS, they can use the same hash function but different internal representation. If you use the implementation specific value, the identical node is going to have different ID in those two CAS so you can't overlay or simple transfer between them.

**Review by cachemeifyoucan — COMMENTED** [review]
Also, we need to update the size calculation for cas-objects right?

File: llvm/lib/CAS/BuiltinCAS.cpp (hunk 1 above)
Comment (cachemeifyoucan) [reply llvm/lib/CAS/BuiltinCAS.cpp:358]: Never mind, I read the comment wrong.

**Comment by dexonsmith** [comment]
Survived building a few things with ASan, and the "full demo" seems fine too. Merging.

**Comment by dexonsmith** [comment]
> Also, we need to update the size calculation for cas-objects right?

Yes, that's right... I'll do separately.

**Pull request merged**